lib/preview/previewUtils.js

```diff
@@ -0,0 +1,183 @@
+const { fetchDomains } = require('../../api/domains');
+const { getAccountId, isTrackingAllowed, getAccountConfig } = require('../config');
+const { platform, release } = require('os');
+const { trackUsage } = require('../../api/fileMapper');
+const unAuth = require('../../api/localDevAuth/unauthenticated');
+
+const VALID_PROXY_DOMAIN_SUFFIXES = ['localhost', 'hslocal.net'];
+const HS_PREVIEW_GATE = "cms:localHublPreviews";
+
+const getPortalDomains = async (portalId) => {
+  try {
+    const result = await fetchDomains(portalId);
+    return result;
+  } catch (error) {
+    console.log("There was a problem fetching domains for your portal. You may be missing a scope necessary for this feature.")
+    return [];
+  }
+}
+
+const getPreviewUrl = (sessionInfo, queryParams) => {
+  const { portalId, env, hublet } = sessionInfo;
+
+  return `http://${portalId}.hubspotpreview${
+    env === 'qa' ? 'qa' : ''
+  }-${hublet}.com/_hcms/preview/template/multi?${stringifyQuery(queryParams)}`;
+}
+
+const stringifyQuery = (query) => {
+  return Object.keys(query)
+    .sort()
+    .map(key => `${key}=${query[key]}`)
+    .join('&');
+}
+
+const insertAtEndOfBody = (html, script) => {
+  const insertAt = (baseStr, index, insertStr) => {
+    return `${baseStr.slice(0, index)}${insertStr}${baseStr.slice(index)}`;
+  }
+  const endOfBodyIndex = html.lastIndexOf("</body>");
+  return insertAt(html, endOfBodyIndex, script);
+}
+
+const addRefreshScript = (html) => {
+  const refreshScript = `
+  <script>
+  (() => {
+    const NORMAL_WAIT_MS = 1000;
+    const BACKOFF_RATIO = 2;
+    let nextWait = NORMAL_WAIT_MS;
+    let timeWaited = 0;
+
+    setInterval(() => {
+      timeWaited += NORMAL_WAIT_MS;
+      if (timeWaited !== nextWait) return;
+      timeWaited = 0;
+      const res = fetch(window.location.origin + '/meta')
+        .then(async (res) => {
+          const hsServerState = await res.json();
+          if (hsServerState["REMOTE_FS_IS_DIRTY"]) {
+            location.reload();
+          }
+          nextWait = NORMAL_WAIT_MS;
+        })
+        .catch(err => {
+          nextWait *= BACKOFF_RATIO;
+          console.log('Disconnected from local server... (retrying in ' + nextWait / 1000 + 's)');
+        });
+    }, NORMAL_WAIT_MS);
+  })();
+  </script>
```

Comment:
Not sure if this is possible, but might be worth looking into if you can just write this whole script as a js file and convert it to a string somehow. In any case, feels like a good candidate for being its own file

Reply:
👍 I agree - planning to move it to be returned by a route on the local server and then instead of embedded the whole js with a script we can just link to that route. Planning to do that after migration to local-dev-lib to make sure we have bundling stuff set up properly with that approach